Added hook tests

tamir-cow opened the pull request:
#### Summary

This PR implements the validation logic for hooks through the `check_hooks` function, ensuring that smart contract hooks are executed correctly according to the protocol rules.

**Related PR**: [link](https://github.com/cowprotocol/circuit_breaker/pull/153) - Contains the data fetching layer and hook extraction logic

#### Hook Validation Rules

The validation logic enforces the following rules for hooks:

1. **Pre-Hook Timing**
   - Pre-hooks must execute before the settlement contract pulls user funds
   - Guaranteed by the hook extraction process in the data fetching layer

2. **Post-Hook Timing**
   - Post-hooks must execute after the settlement contract pushes proceeds to users
   - Guaranteed by the hook extraction process in the data fetching layer

3. **Partially Fillable Orders**
   - **Pre-Hooks**: Must execute only on the first fill
   - **Post-Hooks**: Must execute on every fill (first and subsequent)

4. **Hook Execution Validation**
   - A hook is considered properly executed when there exists an internal CALL in the settlement transaction with matching:
     - Target address
     - Calldata
     - Gas limit

#### Implementation Details

The validation process follows these steps:

1. **Fill Status Determination**
   ```python
   is_first_fill = trade.already_executed_amount == 0
   ```
   - Determines if this is the first fill of an order based on the already executed amount

2. **Hook Candidate Presence Check**
   ```python
   has_no_hook_candidates = (
       not hook_candidates.pre_hooks and 
       not hook_candidates.post_hooks
   )
   if has_no_hook_candidates and is_first_fill:
       return False  # Expected hooks but none executed
   ```
   - If an order has hooks defined and it's the first fill, the transaction must contain hook candidates

3. **Pre-Hook Validation**
   ```python
   # Only check pre-hooks for first fills
   if is_first_fill:
       for pre_hook in hooks.pre_hooks:
           if not _check_hook_execution(pre_hook, hook_candidates.pre_hooks):
               return False
   ```
   - For first fills: All declared pre-hooks must be executed
   - For subsequent fills: Pre-hooks should not be checked

4. **Post-Hook Validation**
   ```python
   # Always check post-hooks regardless of fill status
   for post_hook in hooks.post_hooks:
       if not _check_hook_execution(post_hook, hook_candidates.post_hooks):
           return False
   ```
   - All declared post-hooks must be executed on every fill (first and subsequent)

#### Hook Execution Verification

The `_check_hook_execution` function verifies that a hook was properly executed by checking:

```python
def _check_hook_execution(hook: Hook, hook_candidates: list[Hook]) -> bool:
    for hook_candidate in hook_candidates:
        if (
            hook_candidate.target == hook.target
            and hook_candidate.calldata == hook.calldata
            and hook_candidate.gas_limit == hook.gas_limit
        ):
            return True
    return False
```

This ensures that each required hook has a corresponding executed hook with matching parameters.

#### Data Structures

The implementation uses the following key data structures:

- **Hook**: Contains target address, calldata, and gas limit
- **Hooks**: Contains lists of pre-hooks and post-hooks
- **OnchainSettlementData**: Contains hook candidates extracted from transaction trace
- **OffchainSettlementData**: Contains expected hooks from order appData

#### Future Enhancements

The following rules are planned for future implementation:

- **Hook Attempt Verification**: Verify that hooks are attempted even if they revert
- **No Upstream Reverts**: Ensure intermediate calls between settle() and hook execution don't revert
- **Gas Sufficiency**: Verify that available gas forwarded to hook is ≥ specified gas limit

tamir-cow committed:
Added hook tests

fhenneke committed:
fix bug with missing prices

if a native price is missing, a trade will not contribute to the score computation

fhenneke committed:
add test for missing native price case

fhenneke committed:
fix black

fhenneke committed:
add log when score is skipped

fhenneke committed:
use pytest.parametrize for parameters

tamir-cow committed:
Fixed MyPy warnings

tamir-cow committed:
Finished hook tests implementation

tamir-cow committed:
Fixed UTs

fhenneke reviewed (changes requested):
Made one pass over the code. The general approach of adding hook attributes to onchain and offchain data seems good.

I had some issues matching the rules as described in the docstring of `check_hooks()` to the actual code. Maybe one changes to comments are required here. Will have a look at the fetching logic next.

There should be dedicated unit tests for `check_hooks()`.

tamir-cow commented on circuit_breaker_validator/check_tx.py:
```diff
@@ -140,3 +140,104 @@ def check_score(
+       d. The available gas forwarded to the hook CALL is greater or equal than specified gasLimit
+    """
+    # If there are no hooks in the offchain data, return True
+    if not offchain_data.order_hooks or not onchain_data.hook_candidates:
```
Correct

tamir-cow replied:
That's a good point, I will have to get back to you on that.

tamir-cow commented on circuit_breaker_validator/check_tx.py:
```diff
@@ -140,3 +140,104 @@ def check_score(
+                if (
+                    hook_type == "pre"
+                    and hook_candidate.target == pre_hook.target
+                    and hook_candidate.calldata == pre_hook.calldata
```
There is a check for gas limit a few lines below

fhenneke replied:
I see that check but it also does not do what is written in the rules: `4 a. There exists an internal CALL in the settlement transaction with a
            matching triplet: target, gasLimit, calldata`
It is trivial to add the additional check for `gas_limit` here and makes the code easier to map to rules.

As for the check below: It does not seem to make much sense to me as is. It is not checking that enough gas was available. But as we are currently only trying to implement rules 1. - 4 a) and not rules 4 b) - d) we can just leave it out.

tamir-cow commented on circuit_breaker_validator/models.py:
```diff
@@ -232,6 +252,8 @@ class OffchainSettlementData:
     valid_orders: set[HexBytes]
     jit_order_addresses: set[HexBytes]
     native_prices: dict[HexBytes, int]
+    # hooks data
```
How does it affect the execution?

tamir-cow committed:
Addressed CR comments

tamir-cow committed:
Merge branch 'main' into feature/hook-tests

tamir-cow committed:
Pylint fixes

tamir-cow committed:
Added the following rule:
3. Partially fillable orders:
       a. Should execute the pre-hooks on the first fill only
       b. Should execute the post-hooks on every fill

by checking the executed field to determine if first fill or not

tamir-cow committed:
Lint

tamir-cow committed:
Fixed order of operations

fhenneke commented on circuit_breaker_validator/models.py:
```diff
@@ -214,6 +233,9 @@ class OnchainSettlementData:
     trades: list[OnchainTrade]
+    hook_candidates: list[
+        tuple[str, Hook]
+    ]  # Contains candidates for hooks from transaction trace
```
Is there a reason to not use the `Hooks` data structure for this?

fhenneke reviewed (changes requested):
I added a few comment.

Note that the comment [here](https://github.com/cowprotocol/circuit-breaker-validator/pull/9#issuecomment-3554323896) as well as some of the points raise in the [last review](https://github.com/cowprotocol/circuit-breaker-validator/pull/9#pullrequestreview-3482480800) are largely unaddressed.

tamir-cow committed:
Improved documentation and testing

fhenneke commented on circuit_breaker_validator/check_tx.py:
```diff
@@ -140,3 +145,233 @@ def check_score(
+                )
+                return False
+            hook_executed = True
+            break
```
See comment on docstring of `check_hooks()`.

```suggestion
        if (
            hook_candidate.target == hook.target
            and hook_candidate.calldata == hook.calldata
            and hook_candidate.gas_limit == hook.gas_limit
        ):
            # TODO: Check rules 4 b) - d) here.
            hook_executed = True
            break
```

or, for now, even something like
```python
hook_executed = any((
    hook_candidate.target == hook.target
    and hook_candidate.calldata == hook.calldata
    and hook_candidate.gas_limit == hook.gas_limit
) for hook_candidate in candidates)
```

fhenneke commented on tests/test_check_tx.py:
```diff
@@ -485,3 +488,304 @@ def test_check_score(difference, expected_result):
+        }
+        offchain_data.trades = [offchain_trade]
+
+    elif scenario == "subsequent_fill_with_pre_hooks":
```
(See comment in `_check_order_hooks()`)

fhenneke reviewed (changes requested):
I left a few comment.

The most important ones are about checking for gas and checking for not executing hooks.

The design of the hook test also used some strong assumptions on data fetching. These should be removed or documented.

tamir-cow committed:
Addressed CR comments

tamir-cow committed:
Updated UTs

fhenneke commented on circuit_breaker_validator/models.py:
```diff
@@ -232,6 +274,7 @@ class OffchainSettlementData:
     valid_orders: set[HexBytes]
     jit_order_addresses: set[HexBytes]
     native_prices: dict[HexBytes, int]
+    order_hooks: dict[HexBytes, Hooks]  # Contains hooks from appData of executed trades
```
I would remove the comment and add a docstring.

Due to how you implement the test for hooks (iterating over `order_hooks` instead of iterating over `trades` it becomes critical to document here some guarantees for what `order_hooks` must contain and must not contain. This is one of the reasons why iterating over trades seems better to me. The current implementation contains foot guns which is will easily be missed in future changes.

The guarantee is that `order_hooks` must contain exactly the hooks of executed trades. It is not allowed to contain hooks of orders which are not executed.

There should probably be a test on that as well in the code somewhere. Or, alternatively, one just iterates over trades.

fhenneke commented on tests/test_check_tx.py:
```diff
@@ -485,3 +488,304 @@ def test_check_score(difference, expected_result):
+        offchain_data.trades = [offchain_trade]
+
+    elif scenario == "subsequent_fill_with_pre_hooks":
+        # Subsequent fill - pre-hooks should NOT be executed, but they are
```
That comment should probably more descriptive since the expected behavior is that the test succeeds.

fhenneke reviewed (changes requested):
I still think the logic of checking hooks would benefit from a few changes. If we stick to the current implementation, which I do not recommend, there needs to be more explicit docs and also explicit checks for data consisentency.

tamir-cow committed:
Updated check hooks to iterate over trades rather than order hooks

fedgiac commented on circuit_breaker_validator/check_tx.py:
```diff
@@ -140,3 +145,151 @@ def check_score(
+            f"but no hook candidates found in transaction"
+        )
+        return False
+
```
Is this code really needed? The following code checks if the pre-hooks are in the candidate pre-hooks (if first fill) and if post-hooks are in the candidate post-hooks. This then seems redundant?

tamir-cow replied:
I don't believe so, it is added after encountering this case: 4814655aacfd3db9ebbaf16c428f23372aaf66cd39456c0c5fc15c1759d35588 where we have a mismatch in hooks because of subsequent fills, and based on logic we should skip them.

fedgiac replied:
The transaction you link to has order [0xce7aa6aa17e2e5f0e6cf94ff437de07add4b2852d9098c66237210fabd93dfb0fb632268c37856c6efd48738ef60a74680c4fb626b04786c](https://explorer.cow.fi/orders/0xce7aa6aa17e2e5f0e6cf94ff437de07add4b2852d9098c66237210fabd93dfb0fb632268c37856c6efd48738ef60a74680c4fb626b04786c) in it, which has a hook, so it shouldn't match this block because `has_no_hook_candidates` is false.

But my point here is similar to [this](https://github.com/cowprotocol/circuit-breaker-validator/pull/9/files#r2588562286), with the extra note that it's particularly strange that if `is_first_fill` is false, `not hook_candidates.pre_hooks` is true, and `not hook_candidates.post_hooks` is true, then it skips this early exit condition and continues by using empty loop in the following checks.

harisang replied:
i also don't see why we need `L196-206`. And in this case, i think it makes things more confusing

fedgiac commented on circuit_breaker_validator/check_tx.py:
```diff
@@ -140,3 +145,151 @@ def check_score(
+    # If there are no trades, the rule for hooks is automatically satisfied
+    if not offchain_data.trades:
+        return True
+
```
This isn't needed, right? The loop should be empty and then the function returns true.

fhenneke replied:
Those early exits tend to confuse me more than they help as the case excluded is quite naturally treated with an empty for loop.

All other early exits in this PR seem to also not have an effect on the outcome of running the code but only serve the purpose to reduce mental load and, in the future, make bugs less likely.

For the example
```python
    if not offchain_data.trades:
        return True
```
an can still see how it can help and it does not add too much noise (though more noise than I would like).

For the example
```python
    # If there are hooks in the offchain data and it's the first fill,
    # but there aren't hook candidates in onchain data, return False
    has_no_hook_candidates = (
        not hook_candidates.pre_hooks and not hook_candidates.post_hooks
    )
    if has_no_hook_candidates and is_first_fill:
        logger.error(
            f"Transaction hash {tx_hash!r}: "
            f"Hooks defined for order {trade.order_uid!r} "
            f"but no hook candidates found in transaction"
        )
        return False
```
I would argue that the combination of three conditionals and including negations does not decrease mental load for me and adds quite a bit of noise. I can see this increasing the risk of bugs in the future.

fedgiac commented on tests/test_check_tx.py:
```diff
@@ -485,3 +488,308 @@ def test_check_score(difference, expected_result):
+        ("subsequent_fill_no_pre_hooks", True),
+        # Subsequent fill with pre-hooks present (should ideally fail, but passes because
+        # current implementation doesn't validate that pre-hooks are absent on subsequent fills)
+        ("subsequent_fill_with_pre_hooks", True),
```
I disagree, it will always be expected behavior that subsequent filled orders with pre-hooks are allowed to pass.
This is because solvers are always allowed to add arbitrary calls to a settlement. The call then wouldn't count as a hook but just as a custom solver interaction.
Trying to forbid this would be hard: imagine two users who have the same hook (like in the ETH flow, `wrapAll`), the first user has a fill-or-kill order and needs to call that pre-hook, the second user has a partially fillable order that was already partially executed. The two orders can't be executed together because the second order forbids the solver to include the hook for the first.

fedgiac commented on tests/test_check_tx.py:
```diff
@@ -485,3 +488,308 @@ def test_check_score(difference, expected_result):
+        ("subsequent_fill_with_pre_hooks", True),
+        # Multiple hooks executed correctly - should pass
+        ("multiple_hooks", True),
+    ],
```
A suggestion for an extra test case.

```suggestion
        # Multiple orders with the same hook, executed only once - should pass
        ("multiple_orders_same_hook_executed_once", True),
    ],
```

fedgiac reviewed (commented):
The overall logic and the code looks good. I only have small comments and suggestions.

tamir-cow commented on circuit_breaker_validator/check_tx.py:
```diff
@@ -140,3 +145,151 @@ def check_score(
+       b. The hook needs to be attempted, meaning the hook reverting is not violating any rules
+          - NOT IMPLEMENTED: This requires transaction trace analysis to determine if the hook
+            was attempted. Future implementation may require extending Hook data structure with
+            an 'attempted' attribute.
```
Currently, it is not implemented as the comment says. What do you think we should change in the comment?

fedgiac replied:
Resolved by removing the comment.

tamir-cow committed:
Updated tests

tamir-cow committed:
Updated docstring

tamir-cow committed:
Updated docstring

fhenneke commented on circuit_breaker_validator/models.py:
```diff
@@ -205,20 +226,51 @@ def reverse_protocol_fee(self, trade: OnchainTrade) -> OnchainTrade:
+        order_hooks: Dict mapping order_uid to Hooks for that order.
+            May contain entries for orders not in this settlement without causing issues.
+            Hook validation iterates over trades and looks up hooks from this dict, so
+            extra entries are safely ignored.
```
That comment makes sense in the context of the PR discussion, but does not make sense to me in the context of the code itself.
```suggestion
```

tamir-cow committed:
Update circuit_breaker_validator/models.py

Co-authored-by: Felix Henneke <felix.henneke@protonmail.com>

fedgiac reviewed (commented):
The code looks good, the only unaddressed comments are about coding style and can be disregarded. Not sure if I should approve since I don't know how the process work in this codebase, but you can consider this an approval to the best of my knowledge.

tamir-cow committed:
Merge remote-tracking branch 'origin/main' into feature/hook-tests

tamir-cow committed:
Updated pr based on cr notes

tamir-cow committed:
Lint and black

fhenneke commented on circuit_breaker_validator/check_tx.py:
```diff
@@ -140,3 +145,138 @@ def check_score(
+       b. Should execute the post-hooks on every fill
+          - Explicitly checked: Post-hooks are always validated regardless of fill status
+
+    4. Execution of a hook means:
```
Executing a hook twice is not in violation of any rule. We therefore do not check that something is only executed once.

We generally do  not check multiplicity of hook executions. For example, if the same hook is attached to two orders, we only check that the hook is executed at least once.

There can be cases where we want the hook to be executed at least twice in such cases. But there are valid use cases for not doing that (e.g. for ETH flow orders which wrap all available ETH). So in v0 of hooks we do not consider this to be a violation.

fhenneke reviewed (approved):
The validation logic looks good to me. We should start using the version in our testing environment.
